File: 103binaryTreeZigzagLevelOrderTraversal/binaryTreeZigzagLevelOrderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        final var result = new ArrayList<List<Integer>>();
        var level = Arrays.asList(root);
        while (Objects.nonNull(root) && !level.isEmpty())
        {
            result.add(level.stream().map(node -> node.val).collect(Collectors.toList()));
            level = level.stream().flatMap(node -> Stream.of(node.left, node.right)).filter(Objects::nonNull).collect(Collectors.toList());
        }
        boolean direction = true;
        for (List<Integer> $: result)
        {
            if (!direction) Collections.reverse($);
            direction = !direction;
        }
        return result;
    }
}

File: 103binaryTreeZigzagLevelOrderTraversal/binaryTreeZigzagLevelOrderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

const lodash = require('lodash')

Array.prototype.flatMap = function(callBack)
{
    return lodash.flatMap(this, callBack)
}

var zigzagLevelOrder = function(root) {
    const result = []
    let level = [root], direction = true
    while (root && level.length)
    {
        const val = level.map(node => node.val)
        result.push(direction ? val : val.reverse())
        direction = !direction
        level = level.flatMap(node => [node.left, node.right].filter(_ => _))
    }
    return result
};
